lib/spack/spack/fetch_strategy.py

```diff
@@ -734,24 +743,22 @@ def fetch(self):
 
             # Try to be efficient if we're using a new enough git.
             # This checks out only one branch's history
-            if self.git_version > ver('1.7.10'):
-                args.append('--single-branch')
+            if self.git_version >= ver('1.7.10'):
+                if self.all_branches:
+                    args.append('--no-single-branch')
+                else:
+                    args.append('--single-branch')
 
             with temp_cwd():
-                cloned = False
-                # Yet more efficiency, only download a 1-commit deep tree
-                if self.git_version >= ver('1.7.1'):
-                    try:
-                        git(*(args + ['--depth', '1', self.url]))
-                        cloned = True
-                    except spack.error.SpackError as e:
-                        # This will fail with the dumb HTTP transport
-                        # continue and try without depth, cleanup first
-                        tty.debug(e)
-
-                if not cloned:
-                    args.extend([self.url])
-                    git(*args)
+                # Yet more efficiency: only download a 1-commit deep
+                # tree, if the in-use git and protocol permit it.
+                if (not self.full_depth) and \
+                   self.git_version >= ver('1.7.1') and \
+                   (not self.url.startswith('http:')) and (not self.url.startswith('/')):
```

Comment:
Although this is small, could you bundle this logic into a simple function called something like `using_http_protocol`? It could be a static method of `GitFetchStrategy` I think the comment "This will fail with the dumb HTTP transport..." may be useful there to explain why that check is worthwhile.